crates/payload/primitives/src/payload.rs

```diff
@@ -2,32 +2,78 @@ use crate::{MessageValidationKind, PayloadAttributes};
 use alloc::vec::Vec;
 use alloy_eips::eip4895::Withdrawal;
 use alloy_primitives::B256;
-use alloy_rpc_types_engine::ExecutionPayload;
+use alloy_rpc_types_engine::ExecutionData;
+use core::fmt::Debug;
+use serde::{de::DeserializeOwned, Serialize};
 
-/// Either an [`ExecutionPayload`] or a type that implements the [`PayloadAttributes`] trait.
+/// An execution payload.
+pub trait ExecutionPayload:
+    Serialize + DeserializeOwned + Debug + Clone + Send + Sync + 'static
+{
+    /// Returns the parent hash of the block.
+    fn parent_hash(&self) -> B256;
+
+    /// Returns the hash of the block.
+    fn block_hash(&self) -> B256;
+
+    /// Returns the number of the block.
+    fn block_number(&self) -> u64;
+
+    /// Returns the withdrawals for the payload, if it exists.
+    fn withdrawals(&self) -> Option<&Vec<Withdrawal>>;
+
+    /// Return the parent beacon block root for the payload, if it exists.
+    fn parent_beacon_block_root(&self) -> Option<B256>;
+
+    /// Returns the timestamp to be used in the payload.
+    fn timestamp(&self) -> u64;
+}
+
+impl ExecutionPayload for ExecutionData {
+    fn parent_hash(&self) -> B256 {
+        self.payload.parent_hash()
+    }
+
+    fn block_hash(&self) -> B256 {
+        self.payload.block_hash()
+    }
+
+    fn block_number(&self) -> u64 {
+        self.payload.block_number()
+    }
+
+    fn parent_beacon_block_root(&self) -> Option<B256> {
+        self.sidecar.parent_beacon_block_root()
+    }
+
+    fn withdrawals(&self) -> Option<&Vec<Withdrawal>> {
+        self.payload.withdrawals()
+    }
+
+    fn timestamp(&self) -> u64 {
+        self.payload.timestamp()
+    }
+}
+
+/// Either a type that implements the [`ExecutionPayload`] or a type that implements the
+/// [`PayloadAttributes`] trait.
 ///
 /// This is a helper type to unify pre-validation of version specific fields of the engine API.
 #[derive(Debug)]
-pub enum PayloadOrAttributes<'a, Attributes> {
-    /// An [`ExecutionPayload`] and optional parent beacon block root.
-    ExecutionPayload {
-        /// The inner execution payload
-        payload: &'a ExecutionPayload,
-        /// The parent beacon block root
-        parent_beacon_block_root: Option<B256>,
-    },
+pub enum PayloadOrAttributes<'a, Payload: ExecutionPayload, Attributes> {
```

Comment:
```suggestion
pub enum PayloadOrAttributes<'a, Payload, Attributes> {
```
no need to add this trait bound on the type definition since none of the fields is an associated type of this trait. for example if you had some field as `Payload::SomeAssociatedType` then you would need the trait bound here.